crates/c-api/src/component.rs

```diff
@@ -0,0 +1,659 @@
+use anyhow::{bail, ensure, Context, Result};
+use wasmtime::component::{Component, Func, Instance, Linker, Type, Val};
+use wasmtime::{AsContext, AsContextMut};
+
+use crate::{
+    declare_vecs, handle_call_error, handle_result, wasm_byte_vec_t, wasm_config_t, wasm_engine_t,
+    wasm_name_t, wasm_trap_t, wasmtime_error_t, CStoreContextMut, StoreData,
+};
+use std::collections::HashMap;
+use std::{mem, mem::MaybeUninit, ptr, slice};
+
+#[no_mangle]
+pub extern "C" fn wasmtime_config_component_model_set(c: &mut wasm_config_t, enable: bool) {
+    c.config.wasm_component_model(enable);
+}
+
+pub type wasmtime_component_val_string_t = wasm_byte_vec_t;
+
+#[repr(C)]
+#[derive(Clone)]
+pub struct wasmtime_component_val_record_field_t {
+    pub name: wasm_name_t,
+    pub val: wasmtime_component_val_t,
+}
+
+impl Default for wasmtime_component_val_record_field_t {
+    fn default() -> Self {
+        Self {
+            name: Vec::new().into(),
+            val: Default::default(),
+        }
+    }
+}
+
+declare_vecs! {
+    (
+        name: wasmtime_component_val_vec_t,
+        ty: wasmtime_component_val_t,
+        new: wasmtime_component_val_vec_new,
+        empty: wasmtime_component_val_vec_new_empty,
+        uninit: wasmtime_component_val_vec_new_uninitialized,
+        copy: wasmtime_component_val_vec_copy,
+        delete: wasmtime_component_val_vec_delete,
+    )
+    (
+        name: wasmtime_component_val_record_t,
+        ty: wasmtime_component_val_record_field_t,
+        new: wasmtime_component_val_record_field_vec_new,
+        empty: wasmtime_component_val_record_field_vec_new_empty,
+        uninit: wasmtime_component_val_record_field_vec_new_uninitialized,
+        copy: wasmtime_component_val_record_field_vec_copy,
+        delete: wasmtime_component_val_record_field_vec_delete,
+    )
+    (
+        name: wasmtime_component_val_flags_t,
+        ty: u32,
+        new: wasmtime_component_val_flags_new,
+        empty: wasmtime_component_val_flags_new_empty,
+        uninit: wasmtime_component_val_flags_new_uninitialized,
+        copy: wasmtime_component_val_flags_copy,
+        delete: wasmtime_component_val_flags_delete,
+    )
+}
+
+#[repr(C)]
+#[derive(Clone)]
+pub struct wasmtime_component_val_variant_t {
+    pub index: u32,
+    pub val: Option<Box<wasmtime_component_val_t>>,
+}
+
+#[repr(C)]
+#[derive(Clone)]
+pub struct wasmtime_component_val_result_t {
+    pub value: Option<Box<wasmtime_component_val_t>>,
+    pub error: bool,
+}
+
+#[repr(C)]
+#[derive(Clone)]
+pub struct wasmtime_component_val_enum_t {
+    pub discriminant: u32,
```

Comment:
The Rust API uses the string name as the interface to which discriminant is chosen, personally I think it's more ergonomic in the C-API to use the index, and that's the underlying value stored in `wasmtime::component::Val` anyways.